oximeter/instruments/src/kstat/sampler.rs

```diff
@@ -0,0 +1,723 @@
+                                    );
+                                }
+                            }
+                            match reply_tx.send(Ok(())) {
+                                Ok(_) => trace!(self.log, "sent reply"),
+                                Err(e) => error!(
+                                    self.log,
+                                    "failed to send reply";
+                                    "error" => ?e,
+                                )
+                            }
+                        }
+                        Request::TargetStatus { id, reply_tx } => {
+                            trace!(
+                                self.log,
+                                "request for target status";
+                                "id" => ?id,
+                            );
+                            let response = match self.targets.get(&id) {
+                                None => Err(Error::NoSuchTarget),
+                                Some(SampledObject::Kstat(k)) => {
+                                    Ok(TargetStatus::Ok {
+                                        last_collection: k.time_of_last_collection,
+                                    })
+                                }
+                                Some(SampledObject::Expired(e)) => {
+                                    Ok(TargetStatus::Expired {
+                                        reason: e.reason,
+                                        error: e.error.to_string(),
+                                        expired_at: e.expired_at,
+                                    })
+                                }
+                            };
+                            match reply_tx.send(response) {
+                                Ok(_) => trace!(self.log, "sent reply"),
+                                Err(e) => error!(
+                                    self.log,
+                                    "failed to send reply";
+                                    "id" => ?id,
+                                    "error" => ?e,
+                                ),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Add samples for one target to the internal queue.
+    fn sample_one(
+        &mut self,
+        id: TargetId,
+    ) -> Result<Option<Vec<Sample>>, Error> {
+        let ctl = self.ctl.as_ref().ok_or_else(|| Error::NoKstatCtl)?;
+        let Some(maybe_kstat) = self.targets.get_mut(&id) else {
+            return Ok(None);
+        };
+        let SampledObject::Kstat(sampled_kstat) = maybe_kstat else {
+            panic!("Should not be sampling an expired kstat");
+        };
+
+        let kstats = ctl
+            .iter()
+            .filter(|kstat| sampled_kstat.target.interested(kstat))
+            .zip(sampled_kstat.creation_times.iter())
```

Comment:
I'm worried about the math here. IIUC, the number of elements returned from `ctl.iter.filter` depends on the number of `kstat`s that meet the chosen registrant's criteria *right now*, i.e., at sampling time. But `creation_times` is initialized once in `add_target`, when there might have been a different number of stats. If the former length exceeds the latter then this won't call `read` on the "extra" elements. Is there something else guaranteeing these are equal that I'm overlooking?

(The new tests in `link.rs` only create one test etherstub apiece and always do it before adding any targets, so I'm not sure they cover this case.)

Reply:
That's a good catch, it's definitely possible to be zipping different numbers of elements here.

I think this hits on one of my difficulties on this PR. Kernel statistics publish their creation time as an `hrtime_t`, which is really only useful as a relative timestamp. I do some work to convert it into UTC [here](https://github.com/oxidecomputer/omicron/blob/b6eb2dc9567d202885d73227d916b811821d7b7e/oximeter/instruments/src/kstat/mod.rs#L269), which I am frankly not a huge fan of. That is _intentionally_ done once at registration time, because there's enough slop in the conversions that it starts to matter. Cumulative counters always include a _start time_, and that'll move around by a lot, up to a few milliseconds, in my experimentation. There was actually an illumos issue to report the boot time in UTC, but I'm struggling to find it now. Maybe @jclulow remembers where that is, since I believe he filed it?

Registering them once was my tradeoff, but you're correct that it's wrong! I'll add a regression test for this first off. Now what to do about it? One, we could map directly from the interested kstats to the creation times, which might be a bit tricky. Second would be to compute the creation time on the fly each time. I could just be doing the math wrong, but I worry about that wandering around a bunch and confusing anyone looking at it downstream.

One other idea that occurs to me: we could pretend to start recording the value of the kstat when it's registered (or we see _interest_ in it), and compute the cumulative values over that time interval. That would basically mean we ignore history prior to the registration / interest of the kstat, but that we can just compute a `Utc::now()` all the time.

I think another possibility is to call `KstatTarget::interested` only once, at registration time, not sampling time. I'm sure there are others, but I'll need to stew on the tradeoffs.

Thanks for flagging this, definitely an issue.